kyo-core/shared/src/main/scala/kyo/tags.scala

```diff
@@ -1,41 +1,206 @@
+                                case -1  => false
+                                case idx => loop(idx, superStart, superStart)
+                        end if
+                    }
+                end if
+            end loop
+            loop(subIdx, superIdx, superIdx)
+        end checkSegment
+
+        def nextSegmentEntry(tag: String, idx: Int): Int =
+            @tailrec def loop(idx: Int): Int =
+                if idx >= tag.length() then -1
+                else
+                    tag.charAt(idx) match
+                        case ';'             => idx + 1
+                        case '[' | ',' | ']' => -1
+                        case _               => loop(idx + 1)
+            loop(idx)
+        end nextSegmentEntry
+
+        def sameType(subTag: String, superTag: String, subIdx: Int, superIdx: Int): Boolean =
+            @tailrec def loop(subIdx: Int, superIdx: Int): Boolean =
+                val subChar   = subTag.charAt(subIdx)
+                val superChar = superTag.charAt(superIdx)
+                subChar == superChar && (subChar == ';' || loop(subIdx + 1, superIdx + 1))
+            end loop
+            loop(subIdx, superIdx) && checkParams(subTag, superTag, subIdx, superIdx)
+        end sameType
+
+        def checkParams(subTag: String, superTag: String, subIdx: Int, superIdx: Int): Boolean =
+            val subStart   = subTag.indexOf('[', subIdx)
+            val superStart = subTag.indexOf('[', superIdx)
+            if subStart == -1 || superStart == -1 then
+                superStart == superStart
+            else
+                @tailrec def loop(subIdx: Int, superIdx: Int): Boolean =
+                    val variance = subTag.charAt(subIdx)
+                    variance == superTag.charAt(superIdx) && {
+                        variance match
+                            case '=' =>
+                                sameType(subTag, superTag, subIdx + 1, superIdx + 1)
+                            case '+' =>
+                                checkType(subTag, superTag, subIdx + 1, superIdx + 1)
+                            case '-' =>
+                                checkType(superTag, subTag, superIdx + 1, subIdx + 1)
+                        end match
+                    } && {
+                        val nextSubParam   = nextParam(subTag, subIdx + 1)
+                        val nextSuperParam = nextParam(superTag, superIdx + 1)
+                        if nextSubParam == -1 || nextSuperParam == -1 then
+                            nextSubParam == nextSuperParam
+                        else
+                            loop(nextSubParam, nextSuperParam)
+                        end if
+                    }
+                end loop
+                loop(subStart + 1, superStart + 1)
+            end if
+        end checkParams
+
+        def nextParam(tag: String, idx: Int): Int =
+            @tailrec def loop(opens: Int, idx: Int): Int =
+                tag.charAt(idx) match
```

Comment:
Thanks! I need to get used to this annotation. I had to avoid the pattern guards to add the annotation but it should still improve performance. I added it in a couple of other places as well.